server/src/main/java/org/elasticsearch/action/admin/indices/validate/query/ValidateQueryResponse.java

```diff
@@ -112,22 +143,58 @@ protected void addCustomXContentFields(XContentBuilder builder, Params params) t
             builder.startArray(EXPLANATIONS_FIELD);
             for (QueryExplanation explanation : getQueryExplanation()) {
                 builder.startObject();
-                if (explanation.getIndex() != null) {
-                    builder.field(INDEX_FIELD, explanation.getIndex());
-                }
-                if(explanation.getShard() >= 0) {
-                    builder.field(SHARD_FIELD, explanation.getShard());
-                }
-                builder.field(VALID_FIELD, explanation.isValid());
-                if (explanation.getError() != null) {
-                    builder.field(ERROR_FIELD, explanation.getError());
-                }
-                if (explanation.getExplanation() != null) {
-                    builder.field(EXPLANATION_FIELD, explanation.getExplanation());
-                }
+                explanation.toXContent(builder, params);
                 builder.endObject();
             }
             builder.endArray();
         }
     }
+
+    public static ValidateQueryResponse fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (o instanceof ValidateQueryResponse) {
+            ValidateQueryResponse other = (ValidateQueryResponse) o;
+            Set<QueryExplanation> queryExplSet = new HashSet<>(getQueryExplanation());
+            // We only compare with the index and the shardId because for every failure this is unique.
+            // Also because it is hard to compare Throwable.
+            Set<Tuple<String, Integer>> shardFailureSet =
```

Comment:
@nik9000 @javanna Since its hard to compare exceptions, in this case, I just check for the index and the shardId. Is this alright?

Reply:
I'd prefer to keep the `equals` method complete and only use it in the test when there aren't exceptions. I dunno what we've done other places though.

I'd also like to do something like `assertThat(exception.getMessage(), containsString("original message"));`.

Reply:
The assert part makes sense. However, I don't completely understand what needs to be done for the `equals`.

Reply:
I mean to make `equals` just call `equals` on the Exceptions. It can certainly do `Objects.equals` for the `null`, but I'd prefer the equals make sure the objects are actually equal. And because we can't assert proper equality when round tripping the exceptions I'd hand write an test for the round tripping with exceptions and leave exceptions out of the `AbstractStreamableTestCase` infrastructure.